### PR TITLE
Remove Extra Model in Optimal Saving Lecture

### LINKED ISSUE #107: remove extra function from opt savings
In https://jax.quantecon.org/opt_savings.html it seems the function `create_consumption_model` can be deleted.

After it's deleted `create_consumption_model_jax` can be renamed to `create_consumption_model`.

CC @HumphreyYang @Smit-create

### PR DESCRIPTION
**Pull request by HumphreyYang**
This PR resolves #107.

### CODE REVIEW

**Comment by HumphreyYang** [comment]
Hi @jstac,

This PR is ready for your review. Please kindly let me know if there is anything I can further improve.

Many thanks.

**Comment by jstac** [comment]
Thanks @HumphreyYang 

So here's something weird.  If I run this exact code on my machine, on a CPU, the first round of output, which is from applying `policy_iteration` and appears before the first graph, terminates after 10 iterations.

But here you get

```python
Concluded loop 1 with error 77.
Concluded loop 2 with error 55.
Concluded loop 3 with error 28.
Concluded loop 4 with error 17.
Concluded loop 5 with error 7.
Concluded loop 6 with error 3.
Concluded loop 7 with error 1.
Concluded loop 8 with error 1.
Concluded loop 9 with error 1.
Concluded loop 10 with error 1.
Concluded loop 11 with error 1.
Concluded loop 12 with error 1.
Concluded loop 13 with error 1.
...
```

which continues until loop 250 and then terminates only because 250 is the `maxiter` bound for `policy_iteration`.

Can others replicate?

**Comment by HumphreyYang** [comment]
> Can others replicate?  

Hi @jstac,

I can replicate this on my machine as well. I had the same experience when writing the lecture on Newton's method. In Newton's lecture, the convergence paths on CPU and GPU are very different when tolerance is low. I think this might be related to the accuracy on x86 CPU and GPU (please find a discussion [here](https://developer.nvidia.com/content/precision-performance-floating-point-and-ieee-754-compliance-nvidia-gpus), but most discussions on this are very old). 

I think in this case we want it to converge before hitting the maximum iteration. Should we tune up the tolerance slightly or increase the max iteration?

**Comment by jstac** [comment]
Thanks for the report @HumphreyYang .

We don't want to increase max iter because policy iteration is stuck very close to the optimum.  If we increase max_iter we get the conclusion that policy iteration is slow, even though it converges to the neighborhood of the optimal policy very quickly.

We don't really want to increase the error tolerance because the algorithm is promised to converge to an exact optimal policy in finitely many steps.  

I still find this very surprising.  Is it really using 64 bit floating point arithmetic??  CC @Smit-create and @mmcky , who might be interested.

Maybe the best is to see if slightly changing the parameters stops this odd behavior.  Could you please try shifting gamma to 2.0, or 1.5.  That will reduce curvature, which might help.  w_min could be increased slightly.  And beta could be changed to 0.99.

**Comment by HumphreyYang** [comment]
Hi John, 

> I still find this very surprising. Is it really using 64 bit floating point arithmetic?? CC @Smit-create and @mmcky , who might be interested.

I just checked the `dtype` for intermediate arrays in the computation, they are `float64` or `int64`.

> Maybe the best is to see if slightly changing the parameters stops this odd behavior. Could you please try shifting gamma to 2.0, or 1.5. That will reduce curvature, which might help. w_min could be increased slightly. And beta could be changed to 0.99.

What works:

Setting gamma to 2.0 helps, generating results [here]( https://colab.research.google.com/gist/HumphreyYang/ad812d019688ae6bf81da62d4bba0788/opt_savings.ipynb) and slightly increasing w_min to 0.05 or 0.1 also generate convergence in around 10 loops.  

What does not change the behavior:

Setting gamma to 1.5 does not change the behavior. Holding all parameters the same as before, changing the beta to 0.99 does not change the behavior.

**Comment by jstac** [comment]
Thanks @HumphreyYang 

> Setting gamma to 2.0 helps, generating results [here](https://colab.research.google.com/gist/HumphreyYang/ad812d019688ae6bf81da62d4bba0788/opt_savings.ipynb)

Great.  For now let's change to this parameterization and then we can think about long term fixes.

(The issue seems to be the infinite derivative of the value function at zero.)

Could you also please, in the final plot, change the values of `m_vals` so that `m` ranges over 1 to 441 in steps of 20?

**Comment by HumphreyYang** [comment]
Many thanks John,

> Great. For now let's change to this parameterization and then we can think about long term fixes.

Sure, I will update this.

> Could you also please, in the final plot, change the values of m_vals so that m ranges over 1 to 441 in steps of 20?

Here is the plot it generates:

image

**Comment by Smit-create** [comment]
> I still find this very surprising. Is it really using 64 bit floating point arithmetic?? CC @Smit-create and @mmcky , who might be interested.

I also came across the same problem some time ago where if we use normal `float32` bits, the convergence leads to many iterations. This was fixed by just using `float64`.

**Comment by mmcky** [comment]
I am surprised there is that much difference when using `float32` and `float64` numbers. It looks like `jax` may default to use `float32` and we may need to enable `float64`

https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision

**Comment by jstac** [comment]
@mmcky @Smit-create --- to be clear, I've also had lots of convergence issues using 32 bit floats.  What @HumphreyYang and I are surprised about is that we get different outcomes on the CPU and GPU, even when the GPU is requested to use 64 bit flots.


> I wonder if
> 
> ```python
> jax.config.update("jax_enable_x64", True)
> ```
> 
> needs to be straight after the `jax` import statement?

I don't think so but I could be wrong.  @HumphreyYang any thoughts?  

BTW I need to get this merged and published by tomorrow.

**Comment by HumphreyYang** [comment]
Many thanks @Smit-create and @mmcky for the suggestions.

> > I wonder if
> 
> > 
> 
> > ```python
> 
> > jax.config.update("jax_enable_x64", True)
> 
> > ```
> 
> > 
> 
> > needs to be straight after the `jax` import statement?
> 
> 
> 
> I don't think so but I could be wrong.  @HumphreyYang any thoughts?  

I think our code is already under 64bits given the types are int64 and float64. 

> BTW I need to get this merged and published by tomorrow.

Hi @jstac, I have updated the code to reflect the gamma value. Please kindly check the latest build and let me know if there is anything else I should update before this is merged.

**Comment by jstac** [comment]
Roger that @HumphreyYang , many thanks.  Merging.
